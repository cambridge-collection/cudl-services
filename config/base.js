var config = {}

//Application settings and paths
config.appDir = '/home/cudl/node/cudl-services2';
config.dataDir = '/mnt/cudl-data';
config.cacheDir = '/mnt/cache';

//MySQL settings for cudl database services 
config.mysqlPool = 10;
config.mysqlHost = 'found-dom01.lib.cam.ac.uk';
config.mysqlUser = 'viewerdevuser';
config.mysqlPass = 'resuvedreweiv';
config.mysqlData = 'viewerdev';

//Image server
config.imageServer = 'http://172.22.83.199/iipsrv/iipsrv.fcgi?iiif=';
module.exports = config;
