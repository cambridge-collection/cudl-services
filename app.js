//Config
config = require('../config/base.js');
users = require('../config/users.js');

//Modules
var express = require('express');
var path = require('path');
var favicon = require('static-favicon');
var logger = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');
var mysql = require('mysql');
passport = require('passport')
var strategy = require('passport-localapikey').Strategy;
var fs = require('fs-extra');

//cache directories
fs.ensureDir(config.cacheDir);
fs.ensureDir(config.cacheDir+'/transcriptions');
fs.ensureDir(config.cacheDir+'/translations');

//Routes
var routes = require('./routes/index.js');
var metadata = require('./routes/metadata.js');
var transcription = require('./routes/transcription.js');
var translation = require('./routes/translation.js');
var membership = require('./routes/membership.js');
var embedded = require('./routes/embedded.js');
var iiif = require('./routes/iiif.js');
var app = express();

//MySQL Connection
connection = mysql.createPool({
	connectionLimit	: 10,
        host     : config.mysqlHost,
        user     : config.mysqlUser,
        password : config.mysqlPass,
	database : config.mysqlData,
});


function findByApiKey(apikey, fn) {
    if (apikey in users) {
      return fn(null, users[apikey]);
    }
  return fn(null, null);
}


passport.use(new strategy(
  function(apikey, done) {
    process.nextTick(function () {
      findByApiKey(apikey, function(err, user) {
        if (err) { return done(err); }
        if (!user) { return done(null, false, { message: 'Unknown apikey : ' + apikey }); }
        return done(null, user);
      })
    });
  }
));


// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'jade');

app.use(favicon(path.join(__dirname, '/public/images/brand/favicon.ico')));
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded());
app.use(cookieParser());
app.use(passport.initialize());
app.use(express.static(path.join(__dirname, 'public')));

//app.use('/', routes);
app.use('/v1/metadata', metadata);
app.use('/v1/transcription',transcription);
app.use('/v1/translation', translation);
app.use('/v1/rdb/membership', membership);
app.use('/v1/embedded', embedded);
app.use('/v1/iiif', iiif);

/// catch 404 and forward to error handler
app.use(function(req, res, next) {
    var err = new Error('Not Found');
    err.status = 404;
    next(err);
});

/// error handlers

// development error handler
// will print stacktrace
if (app.get('env') === 'development') {
    app.use(function(err, req, res, next) {
        res.status(err.status || 500);
        res.render('error', {
            message: err.message,
            error: err
        });
    });
}

// production error handler
// no stacktraces leaked to user
app.use(function(err, req, res, next) {
    res.status(err.status || 500);
    res.render('error', {
        message: err.message,
        error: {}
    });
});

console.log('hmmm');
module.exports = app;
