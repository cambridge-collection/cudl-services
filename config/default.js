var config = {};

//Application settings and paths
config.appDir = '/home/cms93/cudl-services';
config.dataDir = '/mnt/cudl-data';
config.cacheDir = '/mnt/cache';
config.dcpdataDir = '/mnt/dcp-data';
config.user = 'chris';
config.group = 'chris';

//Postgres settings for cudl database services
config.postHost = 'cudl-postgres.cmzjzpssbgnq.eu-west-1.rds.amazonaws.com';
config.postUser = 'cudldev';
config.postPass = 'anwqxJcONu1NjUl';
config.postDatabase = 'viewerdev';

config.users = {};
config.users["1a1c31e8-0c04-11e4-b6a8-bf55e15d8c9d"] = { username: 'cudl-dev', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };
config.users["385784fa-0c04-11e4-b5cc-a3745fee5631"] = { username: 'cudl-live', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };
config.users["436b3396-0c04-11e4-b7ed-0308c2efa56c"] = { username: 'genizah', password: 'secret', email: 'genizah@lib.cam.ac.uk' };
config.users["51ce1fb4-0d91-11e4-ad30-77b11da19985"] = { username: 'huw', password: 'secret', email: 'cudl-admin@lib.cam.ac.uk' };

//Image server
config.imageServer = 'http://172.22.83.199/iipsrv/iipsrv.fcgi?iiif=';
config.darwinXTF = 'http://dcp-dev.lib.cam.ac.uk';

module.exports = config;

// Tagging
config.defaultRemoveRatio = 1/5;
