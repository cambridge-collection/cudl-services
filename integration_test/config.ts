export const connectionDetails = {
  host: process.env.TEST_DB_HOST || 'localhost',
  port: Number(process.env.TEST_DB_PORT) || 5432,
  user: process.env.TEST_DB_USER || 'cudlservices',
  password: process.env.TEST_DB_PASSWORD || 'password',
  database: process.env.TEST_DB_DATABASE || 'cudlservices',
};

export const awsEndpointUrl =
  process.env.TEST_AWS_ENDPOINT_URL || 'http://localhost:4566';
