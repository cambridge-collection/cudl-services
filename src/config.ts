export const CONFIG_ENVAR = 'CUDL_SERVICES_CONFIG';

export function loadConfigFromEnvar(): Config {
  const configModulePath = process.env[CONFIG_ENVAR];
  if (!configModulePath) {
    throw new Error(`\
Configuration not found: envar ${CONFIG_ENVAR} be set to the path of the \
config module to load`);
  }

  return require(configModulePath);
}

export interface Config {
  dataDir: string;
  users: Users;
  darwinXTF: string;
  [index: string]: unknown;
}

export interface Users {
  [apiKey: string]: User;
}

export interface User {
  username: string;
  email: string;
}
