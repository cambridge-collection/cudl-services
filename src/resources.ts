export interface Resource {
  close(): Promise<void>;
}

export abstract class BaseResource implements Resource {
  private _isClosed = false;

  async close(): Promise<void> {
    this._isClosed = true;
  }

  isClosed(): boolean {
    return this._isClosed;
  }

  protected ensureNotClosed() {
    if (this.isClosed()) {
      throw new Error('operation on closed resource');
    }
  }
}

export type ResourceUser<A, B> = (resource: A) => Promise<B> | B;

export async function using<A extends Resource, B>(
  resource: PromiseLike<A> | A,
  user: ResourceUser<A, B>
): Promise<B> {
  let resolvedResource: A | undefined = undefined;
  let closeCalled = false;
  try {
    resolvedResource = await resource;
    const result = await user(resolvedResource);
    closeCalled = true;
    await resolvedResource.close();
    return result;
  } catch (e) {
    if (resolvedResource !== undefined && !closeCalled) {
      await resolvedResource.close();
    }
    throw e;
  }
}

export class ExternalResources<T> extends BaseResource {
  private readonly resources: Resource[];
  readonly value: T;

  constructor(value: T, resources: Resource[]) {
    super();
    this.value = value;
    this.resources = Array.from(resources);
  }

  async close(): Promise<void> {
    await super.close();
    await Promise.all(this.resources.map(r => r.close()));
  }
}
