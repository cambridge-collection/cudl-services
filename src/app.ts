import { XSLTExecutor } from '@lib.cam/xslt-nailgun';
import bodyParser from 'body-parser';
import Debugger from 'debug';
import express, { NextFunction, Request, Response } from 'express';
import passport from 'passport';
import path from 'path';
import { CollectionDAO, PostgresCollectionDAO } from './collections';
import { Config, User, Users } from './config';

import { DAOPool, PostgresDatabasePool } from './db';
import {
  CUDLMetadataRepository,
  DefaultCUDLMetadataRepository,
  LegacyDarwinMetadataRepository,
} from './metadata';
import { BaseResource, ExternalResources, using } from './resources';
import * as darwin from './routes/darwin';
import * as membership from './routes/membership';
import * as metadata from './routes/metadata';
import * as similarity from './routes/similarity';
import * as tags from './routes/tags';
import { PostgresTagsDAO, TagsDAO } from './routes/tags-impl';
import * as transcription from './routes/transcription';
import * as translation from './routes/translation';
import { DefaultXTF, XTF } from './xtf';

const cookieParser = require('cookie-parser');
const favicon = require('serve-favicon');
const logger = require('morgan');
const Strategy = require('passport-accesstoken').Strategy;

const debug = Debugger('cudl-services');

// FIXME: move this to server?
//cache directories
// fs.ensureDirSync(config.cacheDir);
// fs.ensureDirSync(config.cacheDir+'/transcriptions');
// fs.ensureDirSync(config.cacheDir+'/translations');

export interface AppOptions {
  users: Users;
  metadataRepository: CUDLMetadataRepository;
  legacyDarwinMetadataRepository: LegacyDarwinMetadataRepository;
  collectionsDAOPool: DAOPool<CollectionDAO>;
  tagsDAOPool: DAOPool<TagsDAO>;
  darwinXtfUrl: string;
  xtf: XTF;
}

export class App extends BaseResource {
  readonly options: AppOptions;
  readonly expressApp: express.Application;
  protected readonly xsltExecutor: XSLTExecutor;

  constructor(options: AppOptions) {
    super();
    this.options = Object.freeze({ ...options });
    this.xsltExecutor = XSLTExecutor.getInstance();
    this.expressApp = this.createExpressApp();
  }

  static fromConfig(config: Config): ExternalResources<App> {
    const dbPool = PostgresDatabasePool.fromConfig(config);

    return new ExternalResources(
      new App({
        metadataRepository: new DefaultCUDLMetadataRepository(config.dataDir),
        legacyDarwinMetadataRepository: new LegacyDarwinMetadataRepository(
          config.legacyDcpDataDir
        ),
        collectionsDAOPool: PostgresCollectionDAO.createPool(dbPool),
        tagsDAOPool: PostgresTagsDAO.createPool(dbPool),
        users: config.users,
        darwinXtfUrl: config.darwinXTF,
        xtf: new DefaultXTF(config),
      }),
      [dbPool]
    );
  }

  private createExpressApp(): express.Application {
    const app = express();

    passport.use(
      new Strategy((token: string, done: (err: any, user: any) => void) => {
        process.nextTick(() => {
          const user = findByApiKey(this.options.users, token);
          return done(null, user || false);
        });
      })
    );

    // view engine setup
    app.set('views', path.resolve(__dirname, '../views'));
    app.set('view engine', 'pug');

    app.use(
      favicon(path.resolve(__dirname, '../public/images/brand/favicon.ico'))
    );
    app.use(logger('dev'));
    app.use(bodyParser.json());
    app.use(bodyParser.urlencoded({ extended: false }));
    app.use(cookieParser());
    app.use(passport.initialize());
    app.use(express.static(path.resolve(__dirname, '../public')));

    // Middleware to redirect trailing slashes to same URL without trailing slash
    app.use((req, res, next) => {
      if (req.url.substr(-1) === '/' && req.url.length > 1) {
        res.redirect(301, req.url.slice(0, -1));
      } else {
        next();
      }
    });

    app.use(
      '/v1/metadata',
      metadata.getRoutes({
        metadataRepository: this.options.metadataRepository,
      })
    );

    app.use(
      '/v1/rdb/membership',
      membership.getRoutes({
        getItemCollections: async (itemID: string) =>
          using(this.options.collectionsDAOPool.getInstance(), dao =>
            dao.getItemCollections(itemID)
          ),
      })
    );

    app.use('/v1/tags', tags.getRoutes({ daoPool: this.options.tagsDAOPool }));

    app.use(
      '/v1/transcription',
      transcription.getRoutes({
        metadataRepository: this.options.metadataRepository,
        legacyDarwinMetadataRepository: this.options
          .legacyDarwinMetadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/translation',
      translation.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xsltExecutor: this.xsltExecutor,
      })
    );

    app.use(
      '/v1/xtf/similarity',
      similarity.getRoutes({
        metadataRepository: this.options.metadataRepository,
        xtf: this.options.xtf,
      })
    );

    app.use(
      '/v1/darwin',
      passport.authenticate('token', { session: false }),
      darwin.getRoutes({ darwinXtfUrl: this.options.darwinXtfUrl })
    );

    // 404 if no route matched
    app.use((req: Request, res: Response, next: NextFunction) => {
      res.status(404).send('Not Found');
    });

    return app;
  }

  async close(): Promise<void> {
    await Promise.all([super.close(), this.xsltExecutor.close()]);
  }
}

function findByApiKey(users: Users, apiKey: string): User | null {
  if (apiKey in users) {
    return users[apiKey];
  }
  return null;
}
