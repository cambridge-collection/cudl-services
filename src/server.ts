import Debugger from 'debug';
import util from 'util';
import {Config, loadConfig} from './config';
import {BaseResource, using} from './resources';
import {BaseError} from './errors';
import * as http from 'http';
import {applyDefaults} from './util';
import {AddressInfo} from 'net';

const debug = Debugger('cudl-services');

export class ServerError extends BaseError {}

export class Server extends BaseResource {
  private readonly server: http.Server;
  public readonly shutdownShutdownComplete: Promise<void>;

  private constructor(server: http.Server, shutdownComplete: Promise<void>) {
    super();
    this.server = server;
    this.shutdownShutdownComplete = shutdownComplete;
  }

  get address(): string | AddressInfo {
    const addr = this.server.address();
    if (addr === null) {
      // Should not happen as the server is listening by the time Server is constructed
      throw new Error('server.address() returned null');
    }
    return addr;
  }

  async close(): Promise<void> {
    if (this.isClosed()) {
      return;
    }
    await super.close();
    this.server.close();
    return this.shutdownShutdownComplete;
  }

  static async start(options?: {port?: number}): Promise<Server> {
    const {port} = applyDefaults(options || {}, {port: 0});

    let config: Config;
    try {
      config = await loadConfig();
    } catch (e) {
      throw new ServerError({
        message: `Failed to load configuration: ${e}`,
        nested: e,
      });
    }

    let serverCreated: (server: http.Server | PromiseLike<http.Server>) => void;
    const serverPromise = new Promise<http.Server>(resolve => {
      serverCreated = resolve;
    });

    const onServerShutdownComplete = using(
      config.createApplication(),
      async application => {
        application.expressApp.set('port', port);

        const server = application.expressApp.listen(port);
        server.once('listening', () => {
          const msg = `Express server listening on ${util.inspect(
            server?.address()
          )}`;
          (debug.enabled ? debug : console.log)(msg);
          serverCreated(server);
        });

        await new Promise((resolve, reject) => {
          function shutdown() {
            process.stdout.write(' Shutting down gracefully... ');
            server.close(err => {
              if (err) {
                reject(err);
              } else {
                resolve(undefined);
              }
            });
          }

          process.once('SIGINT', shutdown);
          process.once('SIGTERM', shutdown);
          server.once('error', reject);
          server.once('close', resolve);
        });
        console.log('server stopped.');
      }
    );

    // We must wait for both promises when waiting for the server, otherwise we'd ignore
    // errors from the process of creating the server.
    try {
      await Promise.race([serverPromise, onServerShutdownComplete]);
    } catch (e) {
      throw new ServerError({
        message: `Failed to start server: ${e}`,
        nested: e,
      });
    }
    return new Server(await serverPromise, onServerShutdownComplete);
  }
}
