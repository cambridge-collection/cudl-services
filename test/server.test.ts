import {mocked} from 'ts-jest/utils';
import {Config, loadConfig} from '../src/config';
import {Application} from '../src/app';
import express from 'express';
import {Server} from '../src/server';
import {AddressInfo} from 'net';
import getPort from 'get-port';
import {using} from '../src/resources';

jest.mock('../src/config', () => {
  return {
    loadConfig: jest.fn(),
  };
});

global.process.getuid = jest.fn(global.process.getuid);

describe('Server', () => {
  let expressApp: express.Application;
  let app: Application;
  let config: Config;

  beforeEach(() => {
    expressApp = express();
    jest.spyOn(expressApp, 'listen');
    app = {
      expressApp,
      close: jest.fn().mockResolvedValue(undefined),
    };
    config = {
      createApplication: jest.fn().mockResolvedValue(app),
    };

    mocked(loadConfig).mockResolvedValue(config);
  });

  describe('start()', () => {
    test('throws if loadConfig() fails', async () => {
      mocked(loadConfig).mockRejectedValueOnce(new Error('Boom!'));
      await expect(Server.start()).rejects.toThrowErrorMatchingSnapshot();
    });

    test('throws if loadConfig().createApplication() fails', async () => {
      mocked(config.createApplication).mockRejectedValueOnce(
        new Error('Boom!')
      );
      await expect(Server.start()).rejects.toThrowErrorMatchingSnapshot();
    });
  });

  describe('running', () => {
    test('runs Application from loadConfig()', async () => {
      expect.assertions(4);
      await using(Server.start(), async server => {
        expect((server.address as AddressInfo).port).toBeTruthy();
        expect(app.expressApp.listen).toHaveBeenCalled();
        expect(app.close).not.toHaveBeenCalled();
      });

      expect(app.close).toHaveBeenCalled();
    });

    test('listens on specified port', async () => {
      const port = await getPort();
      await using(Server.start({port}), async server => {
        expect((server.address as AddressInfo).port).toBe(port);
      });
    });

    test('throws if server fails to bind', async () => {
      const port = await getPort();

      expect.assertions(1);
      await using(Server.start({port}), async () => {
        // Try to listen on the same port again
        await expect(Server.start({port})).rejects.toThrowError(
          `Failed to start server: Error: listen EADDRINUSE: address already in use :::${port}`
        );
      });
    });

    test('shuts down on close()', async () => {
      jest.useFakeTimers();
      const onShutdown = jest.fn();

      const server = await Server.start();
      server.shutdownShutdownComplete.then(onShutdown);

      jest.advanceTimersByTime(1000);
      expect(onShutdown).not.toHaveBeenCalled();

      jest.advanceTimersByTime(10000);
      expect(onShutdown).not.toHaveBeenCalled();

      await server.close();
      expect(onShutdown).toHaveBeenCalled();
      expect(server.isClosed()).toBeTruthy();
    });
  });
});
