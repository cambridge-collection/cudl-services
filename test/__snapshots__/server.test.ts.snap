// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Server start() throws if loadConfig() fails 1`] = `"Failed to load configuration: Error: Boom!"`;

exports[`Server start() throws if loadConfig().createApplication() fails 1`] = `"Failed to start server: Error: Boom!"`;
